### ISSUE TITLE
Compilation error plv8.cc:220:18: error: ‘oid_hash’ was not declared in this scope

### ISSUE THREAD

**Issue by sanindd** [plv8/plv8js#440]
Hi, I'm trying to collect plv8 v2.3.11 for PostgreSQL 13.3 and get a compile error.
Use: CentOS Linux release 7.9.2009, gcc (GCC) 7.3.1, g++ (GCC) 7.3.1

`cd build/v8 ; env CXXFLAGS=-fPIC CFLAGS=-fPIC ninja -C out.gn/x64.release d8
ninja: Entering directory `out.gn/x64.release'
[1/1] Regenerating ninja files
[1/1] Regenerating ninja files
[1/1] Regenerating ninja files
[1/1] Regenerating ninja files
[1/1] Regenerating ninja files
[1/1] Regenerating ninja files
[1/1] Regenerating ninja files
[665/665] LINK ./d8
sed -e 's/^#undef PLV8_VERSION/#define PLV8_VERSION "2.3.11"/' plv8_config.h.in > plv8_config.h
g++ -Wall -O2 -std=c++11 -fno-rtti  -Ibuild/v8/include -Ibuild/v8 -std=c++11 -I. -I./ -I/usr/pgsql-13/include/server -I/usr/pgsql-13/include/internal  -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include -fPIC -c -o plv8.o plv8.cc
plv8.cc: In function ‘void _PG_init()’:
plv8.cc:220:18: error: ‘oid_hash’ was not declared in this scope
  hash_ctl.hash = oid_hash;
                  ^~~~~~~~
plv8.cc:220:18: note: suggested alternative: ‘uuid_hash’
  hash_ctl.hash = oid_hash;
                  ^~~~~~~~
                  uuid_hash
plv8.cc: In function ‘Datum CallFunction(FunctionCallInfo, plv8_exec_env*, int, plv8_type*, plv8_type*)’:
plv8.cc:597:30: error: ‘struct FunctionCallInfoBaseData’ has no member named ‘arg’; did you mean ‘args’?
    args[i] = ToValue(fcinfo->arg[i], fcinfo->argnull[i], &argtypes[i]);
                              ^~~
                              args
plv8.cc:597:46: error: ‘struct FunctionCallInfoBaseData’ has no member named ‘argnull’; did you mean ‘isnull’?
    args[i] = ToValue(fcinfo->arg[i], fcinfo->argnull[i], &argtypes[i]);
                                              ^~~~~~~
                                              isnull
plv8.cc: In function ‘Datum CallSRFunction(FunctionCallInfo, plv8_exec_env*, int, plv8_type*, plv8_type*)’:
plv8.cc:694:29: error: ‘struct FunctionCallInfoBaseData’ has no member named ‘arg’; did you mean ‘args’?
   args[i] = ToValue(fcinfo->arg[i], fcinfo->argnull[i], &argtypes[i]);
                             ^~~
                             args
plv8.cc:694:45: error: ‘struct FunctionCallInfoBaseData’ has no member named ‘argnull’; did you mean ‘isnull’?
   args[i] = ToValue(fcinfo->arg[i], fcinfo->argnull[i], &argtypes[i]);
                                             ^~~~~~~
                                             isnull
plv8.cc: In function ‘Datum common_pl_call_validator(FunctionCallInfo, Dialect)’:
plv8.cc:888:25: error: ‘OPAQUEOID’ was not declared in this scope
    (proc->prorettype == OPAQUEOID && proc->pronargs == 0))
                         ^~~~~~~~~
plv8.cc:888:25: note: suggested alternative: ‘OPEROID’
    (proc->prorettype == OPAQUEOID && proc->pronargs == 0))
                         ^~~~~~~~~
                         OPEROID
plv8.cc: In function ‘v8::Local<v8::Function> find_js_function(Oid)’:
plv8.cc:1387:21: error: ‘HeapTupleGetOid’ was not declared in this scope
    Oid langtupoid = HeapTupleGetOid(tuple);
                     ^~~~~~~~~~~~~~~
plv8.cc:1387:21: note: suggested alternative: ‘HeapTupleData’
    Oid langtupoid = HeapTupleGetOid(tuple);
                     ^~~~~~~~~~~~~~~
                     HeapTupleData
plv8.cc: In function ‘void GetGlobalContext(v8::Persistent<v8::Context>&)’:
plv8.cc:1514:4: error: ‘FunctionCallInfoData’ was not declared in this scope
    FunctionCallInfoData fake_fcinfo;
    ^~~~~~~~~~~~~~~~~~~~
plv8.cc:1514:4: note: suggested alternative: ‘FunctionCallInfoBaseData’
    FunctionCallInfoData fake_fcinfo;
    ^~~~~~~~~~~~~~~~~~~~
    FunctionCallInfoBaseData
In file included from /usr/pgsql-13/include/server/postgres.h:46:0,
                 from plv8.h:19,
                 from plv8.cc:8:
plv8.cc:1525:13: error: ‘fake_fcinfo’ was not declared in this scope
     MemSet(&fake_fcinfo, 0, sizeof(fake_fcinfo));
             ^
/usr/pgsql-13/include/server/c.h:987:31: note: in definition of macro ‘MemSet’
   void   *_vstart = (void *) (start); \
                               ^~~~~
plv8.cc:1525:13: note: suggested alternative: ‘fmgr_info’
     MemSet(&fake_fcinfo, 0, sizeof(fake_fcinfo));
             ^
/usr/pgsql-13/include/server/c.h:987:31: note: in definition of macro ‘MemSet’
   void   *_vstart = (void *) (start); \
                               ^~~~~
plv8.cc:1527:5: error: ‘fake_fcinfo’ was not declared in this scope
     fake_fcinfo.flinfo = &flinfo;
     ^~~~~~~~~~~
plv8.cc:1527:5: note: suggested alternative: ‘fmgr_info’
     fake_fcinfo.flinfo = &flinfo;
     ^~~~~~~~~~~
     fmgr_info
make: *** [Makefile.shared:90: plv8.o] Error 1`

**Comment by JerrySievert** [comment]
correct.  this will not work.

postgres 13 was not supported until 2.3.15, as the release introduced breaking changes.   switch over to 2.3.15 and you should compile fine.

**Comment by marmor** [comment]
> 

The same errors, exactly, happen again with 3.1.2 (PG 13.3).

**Comment by JerrySievert** [comment]
on which os?

**Comment by marmor** [comment]
> 

Linux.
I'm trying to investigate.
One suspicious thing that I'm checking right now, is that Makefile calls pg_config and gets the correct PG_VERSION_NUM, but doesn't do anything with it, just call Makefile.linux which doesn't have it.
Now I'm trying to check if pg_config.h is included before plv8.cc checks PG_VERSION_NUM
(builds take here a looottt of time...).
I'll update in several minutes.

**Comment by JerrySievert** [comment]
`Makefile.linux` only builds `v8`, not plv8 - you don't need `PG_VERSION_NUM` for `v8` since it has zero interaction with postgres.

when you get there, can you supply the output of the error, as well as the output of `pg_config` ?

**Comment by JerrySievert** [comment]
also, you say "linux", but that's a very wide field - which os and version are you attempting to compile with?

**Comment by marmor** [comment]
> `Makefile.linux` only builds `v8`, not plv8 - you don't need `PG_VERSION_NUM` for `v8` since it has zero interaction with postgres.
> 
> when you get there, can you supply the output of the error, as well as the output of `pg_config` ?

You need it because without it plv8.cc doesn't include common/hashfn.h, and then many things become undefined (e.g. oid_hash).
But the build finished, I'm going to look at it and update.

Edit: sorry, I didn't clean it well, have to re-build...
Edit2: output of pg_config is perfect:

BINDIR = /usr/local/pgsql/bin
DOCDIR = /usr/local/pgsql/share/doc
HTMLDIR = /usr/local/pgsql/share/doc
INCLUDEDIR = /usr/local/pgsql/include
PKGINCLUDEDIR = /usr/local/pgsql/include
INCLUDEDIR-SERVER = /usr/local/pgsql/include/server
LIBDIR = /usr/local/pgsql/lib
PKGLIBDIR = /usr/local/pgsql/lib
LOCALEDIR = /usr/local/pgsql/share/locale
MANDIR = /usr/local/pgsql/share/man
SHAREDIR = /usr/local/pgsql/share
SYSCONFDIR = /usr/local/pgsql/etc
PGXS = /usr/local/pgsql/lib/pgxs/src/makefiles/pgxs.mk
CONFIGURE = 
CC = gcc
CPPFLAGS = -D_GNU_SOURCE
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -O2
CFLAGS_SL = -fPIC
LDFLAGS = -Wl,--as-needed -Wl,-rpath,'/usr/local/pgsql/lib',--enable-new-dtags
LDFLAGS_EX = 
LDFLAGS_SL = 
LIBS = -lpgcommon -lpgport -lpthread -lz -lreadline -lrt -ldl -lm 
VERSION = PostgreSQL 13.3

**Comment by marmor** [comment]
> also, you say "linux", but that's a very wide field - which os and version are you attempting to compile with?



> also, you say "linux", but that's a very wide field - which os and version are you attempting to compile with?

It's a docker container:

cat /etc/os-release 

PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

**Comment by JerrySievert** [comment]
try bullseye, it's supported: https://github.com/plv8/plv8/blob/r3.1/platforms/Docker/bullseye/14.5/Dockerfile

> You need it because without it plv8.cc doesn't include common/hashfn.h, and then many things become undefined (e.g. oid_hash).

`plv8` needs it, `v8` does not.   `Makefile.linux` does not need it because it builds the dependency `v8` and does not need to know about Postgres.  `plv8.cc` needs it because it is part of `plv8` and is not built by `Makefile.linux`.

those makefiles are there for dependency building, not for building `plv8`.

**Comment by marmor** [comment]
> try bullseye, it's supported: https://github.com/plv8/plv8/blob/r3.1/platforms/Docker/bullseye/14.5/Dockerfile
> 
> > You need it because without it plv8.cc doesn't include common/hashfn.h, and then many things become undefined (e.g. oid_hash).
> 
> `plv8` needs it, `v8` does not. `Makefile.linux` does not need it because it builds the dependency `v8` and does not need to know about Postgres. `plv8.cc` needs it because it is part of `plv8` and is not built by `Makefile.linux`.
> 
> those makefiles are there for dependency building, not for building `plv8`.

Oops, sorry, I probably read your comment too fast... ;-)
However, the bottom line is the same:
plv8.cc was compiled without PG_VERSION_NUM, which may explain why it didn't include common/hashfn.h and then failed on many undefineds (e.g. oid_hash) (unless it got PG_VERSION_NUM from pg_config.h which is what I'm trying to find...).

Edit: meanwhile, a different build (pgxn install plv8) finished, failed too (but because of another reason - unable to auto-detect email address).
Back to make...

Edit2: now it fails earlier (strange!) during the build of v8:
./torque: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.26' not found (required by ./torque)

I'll continue to play with it later...

Thanks!

**Comment by JerrySievert** [comment]
if it managed to try to compile without `PG_VERSION_NUM`, then likely there was some failure at the beginning due to `pg_config` not being in the path - that’s what i’d look at first.  if you see errors the include `le` and `ge` at the beginning of the compilation, those are errors trying to deal with an empty result from `pg_config`

**Comment by marmor** [comment]
> `Makefile.linux` only builds `v8`, not plv8 - you don't need `PG_VERSION_NUM` for `v8` since it has zero interaction with postgres.
> 
> when you get there, can you supply the output of the error, as well as the output of `pg_config` ?

Well, everything is OK...
When I started the build, not everything was OK:
Some dependencies were missing, as well as some environment variables (not only PG_CONFIG, but also other variables that without setting them, the build fails, like LD_LIBRARY_PATH=/build/plv8-3.1.2/build/v8/build/linux/debian_sid_amd64-sysroot/usr/lib/x86_64-linux-gnu - libstdc under buster supports all of the GLIBCXX_3.4 versions till 25, but not 26... and can't be upgraded...).
So I built, failed, apt-get install, built again, failed again, export PG_CONFIG=/usr/local/pgsql/bin/pg_config, built again, and so on.
Unfortunately, I didn't make clean between actions...
And I received exactly, but exactly, the same output as the issue opener.
Which made me think that the problem is back...
So I'm sorry and hope I haven't bother you too much...

By the way:
2 things attracted me to plv8, and both of them are not relevant: 😉

1. JS-based FDW:
Well, as "trusted", it's not possible to do any of the actions that each FDW is required to do...
(like connecting to HTTP resources, accessing the FS, etc.)

2. Verification of *asymmetric* (RSA) JWTs:
Right now PostgreSQL supports only symmetric and not RSA.
However, I'm afraid that the price adding plv8 just for this purpose, is too expensive
(I think that it will be much cheaper to do in Python;
Customers dream about getting a smaller container from me, while plv8 will cause my container to be bigger...).

Yet, the option to run JS from within PG, looks too cool to avoid... 😉

So in any case, I want to thank you and everybody who helped this project!
